File: Python/leetcode/shift-2d-grid/solution.py
from typing import List


class Solution:
    def shiftGrid(self, grid: List[List[int]], k: int) -> List[List[int]]:
        m = len(grid)
        if m == 0:
            return grid
        n = len(grid[0])
        if n == 0:
            return grid
        total = m * n
        k %= total
        flat = [grid[i][j] for i in range(m) for j in range(n)]
        res = [[0] * n for _ in range(m)]
        for idx in range(total):
            np = (idx + k) % total
            res[np // n][np % n] = flat[idx]
        return res

File: Rust/leetcode/shift-2d-grid/solution.rs
impl Solution {
    pub fn shift_grid(grid: Vec<Vec<i32>>, k: i32) -> Vec<Vec<i32>> {
        let m = grid.len();
        if m == 0 {
            return grid;
        }
        let n = grid[0].len();
        if n == 0 {
            return grid;
        }
        let total = m * n;
        let k = (k as usize) % total;
        let mut flat = Vec::with_capacity(total);
        for row in &grid {
            for &num in row {
                flat.push(num);
            }
        }
        let mut res = vec![vec![0i32; n]; m];
        for idx in 0..total {
            let np = (idx + k) % total;
            res[np / n][np % n] = flat[idx];
        }
        res
    }
}
